### DISCUSSION TITLE
reacting to the view : get current view data

### DISCUSSION THREAD

**Discussion by BUGHERE** [GMOD/jbrowse-components#3602] [Q&A] [answered]
I am using JBrowse2 to show my bed data like this:
![image](https://user-images.githubusercontent.com/55886903/226564712-a41415b3-180f-4bd4-b6d0-e1ef743cac80.png)
**Is there any method to get all feature details data(grey blocks in the img) in current view? Such as, I want to get all lengthes of all grey blocks**

By the way, I can only get the feature detail when I click it manually, and using onChange method i can get current feature detail data(like https://jbrowse.org/jb2/docs/tutorials/embed_linear_genome_view/08_reacting/):
`onChange: patch => {
  updates.innerHTML += JSON.stringify(patch) + '\n'
},`
![image](https://user-images.githubusercontent.com/55886903/226565423-ffe038a7-274e-4c03-bba6-4e997aee27f4.png)

If any one can help me, I will appreciate it.

**Comment by cmdcolin** [comment, marked as answer]
this is an interesting question


>By the way, I can only get the feature detail when I click it manually, and using onChange method i can get current feature detail data(like https://jbrowse.org/jb2/docs/tutorials/embed_linear_genome_view/08_reacting/):

the feature can probably be obtained with that onClick, or can be reacted to by "observing" the "session.selection" object. so for example if you had

```
const DisplaySelectedFeature = observer(({session}) => {
    return <div>{JSON.stringify(session.selection)}</div>
})
```

this will automatically react to the currently selected feature (the observer component automatically re-renders when session.selection changes)

> Is there any method to get all feature details data(grey blocks in the img) in current view? Such as, I want to get all lengthes of all grey blocks

currently, this is actually difficult to do. it used to be a bit easier but the alignments track code actually attempts to specifically NOT make feature data available. 

 this sounds odd, but it is part of our effort to make our webworker code faster.  the background to this issue is that jbrowse can use webworkers and this creates a hard 'boundary' where memory cannot be freely shared across the boundary. previously, we serialized all the feature information from the web worker to the main thread, but stopped doing this, instead only requesting information from the webworker on demand. this was part of the 'noserialize' change mentioned here, it made our code when it uses webworkers about 2x faster, alongside decreasing memory consumption, etc making it more scalable https://jbrowse.org/jb2/blog/2022/04/26/v1.7.5-release/ 

so, in order to do what you are asking e.g. get information about all features available in the view, you may need to actually fetch them manually

I tried making a demo in our storybook to demonstrate 

https://github.com/GMOD/jbrowse-components/compare/load_visible_features

live https://jbrowse.org/storybook/lgv/load_visible_features/?path=/story/linear-view--with-show-track

**Reply by cmdcolin** [reply]
note that by doing the below, you will actually be serializing all the features and can be memory intensive on large BAMs

**Reply by BUGHERE** [reply]
Thank you very much! This helps me a lot.
